specs/_features/eip6110/beacon-chain.md

```diff
@@ -0,0 +1,413 @@
+    process_sync_aggregate(state, block.body.sync_aggregate)
+```
+
+#### Modified `process_operations`
+
+*Note*: The function `process_operations` is modified to process `DepositReceipt` operations included in the payload.
+
+```python
+def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
+    # [New in EIP-6110]
+    # Prevent potential underflow introduced by mixing two deposit processing flows
+    if state.eth1_data.deposit_count > state.eth1_deposit_index:
+        assert len(body.deposits) == min(MAX_DEPOSITS, state.eth1_data.deposit_count - state.eth1_deposit_index)
+    else:
+        assert len(body.deposits) == 0
+
+    def for_ops(operations: Sequence[Any], fn: Callable[[BeaconState, Any], None]) -> None:
+        for operation in operations:
+            fn(state, operation)
+
+    for_ops(body.proposer_slashings, process_proposer_slashing)
+    for_ops(body.attester_slashings, process_attester_slashing)
+    for_ops(body.attestations, process_attestation)
+    for_ops(body.deposits, process_deposit)  # [Modified in EIP-6110]
+    for_ops(body.voluntary_exits, process_voluntary_exit)
+    for_ops(body.bls_to_execution_changes, process_bls_to_execution_change)
+
+    # [New in EIP-6110]
+    if is_execution_enabled(state, body):
+        for_ops(body.execution_payload.deposit_receipts, process_deposit_receipt)
+```
+
+#### New `get_validator_from_deposit_data`
+
+```python
+def get_validator_from_deposit_data(pubkey: BLSPubkey, withdrawal_credentials: Bytes32, amount: uint64) -> Validator:
+    effective_balance = min(amount - amount % EFFECTIVE_BALANCE_INCREMENT, MAX_EFFECTIVE_BALANCE)
+
+    return Validator(
+        pubkey=pubkey,
+        withdrawal_credentials=withdrawal_credentials,
+        activation_eligibility_epoch=FAR_FUTURE_EPOCH,
+        activation_epoch=FAR_FUTURE_EPOCH,
+        exit_epoch=FAR_FUTURE_EPOCH,
+        withdrawable_epoch=FAR_FUTURE_EPOCH,
+        effective_balance=effective_balance,
+    )
+```
+
+#### New `apply_deposit`
+
+```python
+def apply_deposit(state: BeaconState,
+                  pubkey: BLSPubkey,
+                  withdrawal_credentials: Bytes32,
+                  amount: uint64,
+                  signature: BLSSignature,
+                  ) -> None:
+    validator_pubkeys = [validator.pubkey for validator in state.validators]
+    if pubkey not in validator_pubkeys:
+        # Verify the deposit signature (proof of possession) which is not checked by the deposit contract
+        deposit_message = DepositMessage(
+            pubkey=pubkey,
+            withdrawal_credentials=withdrawal_credentials,
+            amount=amount,
+        )
+        domain = compute_domain(DOMAIN_DEPOSIT)  # Fork-agnostic domain since deposits are valid across forks
+        signing_root = compute_signing_root(deposit_message, domain)
+        # Initialize validator if the deposit signature is valid
+        if bls.Verify(pubkey, signing_root, signature):
```

Comment:
I'd prefer to keep this as parallel to the previous fork `process_deposit` functions as possible -- so I'd prefer to use the `if not` gating condition.

Also, we could make a new helper for this and past phases  -- `verify_and_add_validator()` that handles all of this logic (from line 247 to 260 here).

Reply:
In fact, I guess we could make `apply_deposit` the generic function, define it starting in phase 0, and just utilize it here

Reply:
Yeah, `apply_deposit` was introduced here for deduplication purposes, it has the same logic as `process_deposit` starting from the point after merkle proof verification and bumping deposit index in the state.

Probably, we can introduce this function in Altair spec and reuse it here. If we did this for the Phase0 then it would have to be overridden in Altair.

Reply:
which isn't the worst. It's a nice consistency throughout even with the altair override